src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java

```diff
@@ -591,12 +602,13 @@ public boolean equals(Object o) {
       return Arrays.equals(digest, that.digest)
           && size == that.size
           && locationIndex == that.locationIndex
-          && Objects.equals(actionId, that.actionId);
+          && Objects.equals(actionId, that.actionId)
+          && expiredAtEpochMilli == that.expiredAtEpochMilli;
```

Comment:
1. Does that increase the size of this object? If so, it may be a good time to split actionId and this timestamp to a new subclass of its own.
2. Can an action rerun renew this? If so, will that not negatively affect downstream action through no skyframe pruning?

Reply:
> 1. Does that increase the size of this object? If so, it may be a good time to split actionId and this timestamp to a new subclass of its own.

We plan to remove `actionId`. I can do it in a separate PR. Does that sound good to you? I am open to create a subclass too.



> 2\. Can an action rerun renew this? If so, will that not negatively affect downstream action through no skyframe pruning?

Yes. Imaging the metadata expired in an incremental build, if the file doesn't exist locally, skyframe will invalidate the action node and re-evaluate it. The action cache will be skipped because we also check the expiration there. So the generating action will be re-executed, hitting the remote cache and inject a remote metadata with new expiration.

Reply:
> We plan to remove `actionId`. I can do it in a separate PR. Does that sound good to you? I am open to create a subclass too.

My take is that I would strongly recommend running one of the large builds upon import and checking memory impact.

> So the generating action will be re-executed, hitting the remote cache and inject a remote metadata with new expiration.

Let me get back to my question. After such execution, with the same results, we will not evaluate the `ActionExecutionValue` as equal for downstream actions due your newly added condition. Therefore, change-pruning will not kick in, even though it could. You will get an action cache hit for downstream actions, but it should be unnecessary to check that at all for those had change-pruning worked.

Reply:
That's true. Maybe exclude `expiration` from `equals`?